File: api-js/src/controllers/graficaController.js
const pool = require('../config/db.js');

const procedimientosPermitidos = [
  'cantidadAlumnosPorTipoArbol',
  'cantidadMonitoreosConPlagasDetectadas',
  'cantidadArbolesPorMetodoRiego',
  'cantidadArbolesPorTipo',
  'modaAltura',
  'modaTipoArboles',
  'promedioAltura',
  'promedioDiametroPorTipo',
  'promedioPlantadosPorMes',
  'cantidadRiegosPorArbol'
];

const obtenerDatos = (procedure, res) => {
  if (!procedimientosPermitidos.includes(procedure)) {
    return res.status(400).json({ error: 'Procedimiento no permitido' });
  }
  pool.query(`CALL ${procedure}()`)
    .then(([rows]) => {
      res.json(rows[0] || rows);
    })
    .catch(err => {
      console.error(err);
      res.status(500).json({ error: 'Error en la consulta' });
    });
};

module.exports = {
  alumnosPorArbol: (req, res) => obtenerDatos('cantidadAlumnosPorTipoArbol', res),
  cantidadPlagas: (req, res) => obtenerDatos('cantidadMonitoreosConPlagasDetectadas', res),
  arbolesPorRiego: (req, res) => obtenerDatos('cantidadArbolesPorMetodoRiego', res),
  cantidadArbolesTipo: (req, res) => obtenerDatos('cantidadArbolesPorTipo', res),
  modaAltura: (req, res) => obtenerDatos('modaAltura', res),
  modaTipoArbol: (req, res) => obtenerDatos('modaTipoArboles', res),
  promedioAltura: (req, res) => obtenerDatos('promedioAltura', res),
  promedioDiametro: (req, res) => obtenerDatos('promedioDiametroPorTipo', res),
  promedioArbolesMes: (req, res) => obtenerDatos('promedioPlantadosPorMes', res),
  riegoPorArbol: (req, res) => obtenerDatos('cantidadRiegosPorArbol', res)
};

File: api-js/app/controllers/grafica_controller.py
from flask import Blueprint, jsonify

from app.config.db import get_db_connection

grafica_bp = Blueprint('grafica_bp', __name__)

def ejecutar_procedimiento(nombre):
    conn = get_db_connection()
    if not conn:
        return jsonify({'error': 'No se pudo conectar a la base de datos'}), 500
    try:
        cursor = conn.cursor(dictionary=True)
        cursor.callproc(nombre)
        resultados = []
        for result in cursor.stored_results():
            resultados = result.fetchall()
        cursor.close()
        conn.close()
        return jsonify(resultados)
    except Exception as e:
        print("Error en la consulta:", e)
        return jsonify({'error': 'Error en la consulta'}), 500

def alumnos_por_arbol():
    return ejecutar_procedimiento('cantidadAlumnosPorTipoArbol')

def cantidad_plagas():
    return ejecutar_procedimiento('cantidadMonitoreosConPlagasDetectadas')

def arboles_por_riego():
    return ejecutar_procedimiento('cantidadArbolesPorMetodoRiego')

def cantidad_arboles_tipo():
    return ejecutar_procedimiento('cantidadArbolesPorTipo')

def moda_altura():
    return ejecutar_procedimiento('modaAltura')

def moda_tipo_arbol():
    return ejecutar_procedimiento('modaTipoArboles')

def promedio_altura():
    return ejecutar_procedimiento('promedioAltura')

def promedio_diametro():
    return ejecutar_procedimiento('promedioDiametroPorTipo')

def promedio_arboles_mes():
    return ejecutar_procedimiento('promedioPlantadosPorMes')

def riego_por_arbol():
    return ejecutar_procedimiento('cantidadRiegosPorArbol')
